mlflow/sklearn/__init__.py

```diff
@@ -1193,10 +1240,38 @@ def fit_mlflow(original, self, *args, **kwargs):
         autologging_client = MlflowAutologgingQueueingClient()
         _log_pretraining_metadata(autologging_client, self, *args, **kwargs)
         params_logging_future = autologging_client.flush(synchronous=False)
-        fit_output = original(self, *args, **kwargs)
+
+        if flavor_name == "xgboost_sklearn":
+            import mlflow.xgboost
+
+            # mlflow xgboost autologging items:
+            # (1) record eval results and (2) log feature importance plot
+            if self.importance_type is None:
+                importance_types = ["weight"]
+            else:
+                importance_types = (
+                    self.importance_type
+                    if isinstance(self.importance_type, list)
+                    else [self.importance_type]
+                )
+
+            (
+                fit_output,
+                early_stopping,
+                early_stopping_logging_operations,
+            ) = mlflow.xgboost._mlflow_xgboost_logging(
+                importance_types, autologging_client, _logger, original, self, *args, **kwargs,
+            )
+        else:
+            fit_output = original(self, *args, **kwargs)
+
         _log_posttraining_metadata(autologging_client, self, *args, **kwargs)
         autologging_client.flush(synchronous=True)
         params_logging_future.await_completion()
+
+        if flavor_name == "xgboost_sklearn" and early_stopping:
+            early_stopping_logging_operations.await_completion()
```

Comment:
Do we need to combine sklearn autologging content with xgboost autologging content? Assuming that `_mlflow_xgboost_logging` logs a comprehensive set of parameters, training metrics, and artifacts, it seems sufficient to only call `_mlflow_xgboost_logging` for XGBoost scikit-learn estimators.

In that case, can we move this logic out of `fit_mlflow()` and instead add a new `fit_impl()` argument to `patched_fit()`? https://github.com/mlflow/mlflow/blob/e919f024cc1c282b32f0951e446ce5d0d3c833f7/mlflow/sklearn/__init__.py#L1361-L1389. For XGBoost scikit-learn models, we can pass in the `train()` method defined in `mlflow.xgboost.autolog(), and for other scikit-learn models we can pass in `fit_mlflow` (which we should to rename to `fit_mlflow_sklearn`).

This way, we also don't need to introduce a separate `_mlflow_xgboost_logging` method.

Reply:
@dbczumar Thank you for your review! Maybe I misunderstood our discussion [1](https://github.com/mlflow/mlflow/issues/4296#issuecomment-938374819), [2](https://github.com/mlflow/mlflow/issues/4296#issuecomment-938742898), [3](https://github.com/mlflow/mlflow/issues/4296#issuecomment-939160361): I thought we would like to keep `_log_pretraining_metadata()` and `_log_posttraining_metadata()` in `fit_mlflow()` when logging XGBoost sklearn models. 

Simply adding 
```python
safe_patch(FLAVOR_NAME, xgboost.sklearn, "train", train, True)
```
to `mlflow.xgboost.autolog()` enables the autologging for XGBoost sklearn models, although it won't use the `mlflow.sklearn` routine.

**Pros:** In this way, we don't have to add call backs and feature importance plot to sklearn fit methods. 

**Cons:** The `train()` method returns a Booster object. If `train()` is patched, the logged `model_class` would be a Booster not the actual XGBoost sklearn model.

Do you suggest patching `xgboost.sklearn.train()` using `patch_fit()` (in `mlflow.sklearn`), which still take advantage of other `mlflow.sklearn` autologging functionalities. Maybe we can pass a variable indicating the model class as well. 


Please correct me if I misunderstood your ideas. Thanks!

Reply:
In that case, can we do the following:
* patch `xgboost.sklearn.train()` in `mlflow.xgboost`
* patch other methods, such as `patched_predict`, in `mlflow.sklearn` for XGBoost sklearn models but **not** for `fit_mlflow()`.

Reply:
> **Cons:** The `train()` method returns a Booster object. If `train()` is patched, the logged `model_class` would be a Booster not the actual XGBoost sklearn model.

Hi @dbczumar, I think it is a way to resolve this issue. We inspect the caller info insider the `train()` method:
```python
# in mlflow.xgboost
def autolog():
    # ...
    def train():
        # ...
        def _get_xgb_caller_info():
            import inspect
            xgb_caller = inspect.stack()[4]
            is_caller_fit = xgb_caller[3] == "fit"
            if not is_caller_fit: return "xgboost"
            else:
                caller_class = xgb_caller[0].f_locals["self"].__class__
                return caller_class.__module__ + "." + caller_class.__name__
        xgb_caller = _get_xgb_caller_info()
        # current implementation
```
If `xgb_caller` is `xgboost`, we save / log a Booster object. Otherwise, we save / log the model based on its class info. 

Please let me know if these plans look good to you, i.e., (1) patch `train()` to `mlflow.sklearn` without modifying `fit_mlflow()` routine; (2) activate other sklearn patched method for `mlflow.xgboost` but no `fit`. Thanks!